### PR TITLE
VDIV, VDIVE and VMOD instructions

### PR DESCRIPTION
**Pull request by alqotel**

### CODE REVIEW

**Review by mferst — COMMENTED** [review]
There are also some style errors. You can run checkpatch.pl on a source file without generating the patch. E.g.: `./scripts/checkpatch.pl target/ppc/int_helper.c`

File: target/ppc/translate/vmx-impl.c.inc [hunk 1]
```diff
@@ -2104,6 +2104,134 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+    tcg_temp_free_i32(l);
+    tcg_temp_free_i64(val1);
+    tcg_temp_free_i64(val2);
+}
+
+#define REM_VEC(NAME, TL, REM)                                  \
+static void do_##NAME (TCGv_##TL t, TCGv_##TL a, TCGv_##TL b)   \
+{                                                               \
+    TCGv_##TL temp;                                             \
+    temp = tcg_temp_local_new_##TL();                           \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vmx-impl.c.inc:2209]: This one is incorrect, I forgot to change when I stopped using brcond

File: target/ppc/translate/vmx-impl.c.inc [hunk 2]
```diff
@@ -2104,6 +2104,134 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+    rd = gen_avr_ptr(a->vrt);
+    gen_helper(rd, ra, rb);
+    tcg_temp_free_ptr(ra);
+    tcg_temp_free_ptr(rb);
+    tcg_temp_free_ptr(rd);
+
+    return true;
+}
+
+#define TRANS_VX(FLAGS, NAME, VECE, FNI4_FUNC, FNI8_FUNC)               \
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vmx-impl.c.inc:2124]: Any suggestions? Maybe be really direct with a name like `TRANS_VX_GVEC_32_64`?

[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vmx-impl.c.inc:2124]: I'm not sure if we'll have other VX-form insns with the same requirements to reuse this macro, so I'd use the insn name. E.g. `TRANS_VDIV` first and maybe rename it to `TRANS_VDIV_VMOD` later.

File: target/ppc/translate/vmx-impl.c.inc [hunk 3]
```diff
@@ -2104,6 +2104,134 @@ static bool trans_VPEXTD(DisasContext *ctx, arg_VX *a)
...
+    tcg_gen_concat_i32_i64(val1, tcg_constant_i32(0), a);
+    tcg_gen_ext_i32_i64(val2, b);
+    tcg_gen_div_i64(res, val1, val2); // (a << 32)/b
+    tcg_gen_andi_i64(val2, res, 0xFFFFFFFF8ull << 28);
+    tcg_gen_movcond_i64(TCG_COND_EQ, val1, val2, tcg_constant_i64(0), res,
+                        tcg_constant_i64(0));
+    tcg_gen_movcond_i64(TCG_COND_EQ, val2, val2,
+                        tcg_constant_i64(0xFFFFFFFF8ull << 28), res,
+                        tcg_constant_i64(0));
+    tcg_gen_or_i64(res, val1, val2);
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vmx-impl.c.inc:2173]: Good idea, but wouldn't in be better to do a `tcg_gen_ext32s_i64(val1, res)` in that case? Make it clearer that we're just extending the signal bit to all bits in the upper half

[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/translate/vmx-impl.c.inc:2173]: Yes, it's better.

File: target/ppc/translate/vmx-impl.c.inc [hunk 4]
```diff
@@ -2200,13 +2200,38 @@ static void do_diveu_i32(TCGv_i32 t, TCGv_i32 a, TCGv_i32 b)
     tcg_temp_free_i64(val2);
 }
 
+#define REM_VEC(NAME, SZ, REM)                                  \
+static void do_##NAME(TCGv_##SZ t, TCGv_##SZ a, TCGv_##SZ b)    \
+{                                                               \
+    TCGv_##SZ temp;                                             \
+    TCGv_##SZ zero = tcg_constant_##SZ(0);                      \
+    temp = tcg_temp_new_##SZ();                                 \
+    REM(temp, a, b);                                            \
```
Comment (mferst) [inline target/ppc/translate/vmx-impl.c.inc:2209]: If `b` is zero, the instruction used by the backend might deliver a signal to the process. E.g., `tcg_gen_rem_i64` emits an idiv in x86.

File: target/ppc/int_helper.c [hunk 5]
```diff
@@ -1227,6 +1227,24 @@ void helper_VDIVEUQ(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
     }
 }
 
+void helper_VMODSQ(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
+{
+    if (likely(int128_nz(b->s128))) {
+        t->s128 = int128_rems(a->s128, b->s128);
+    } else {
+        t->s128 = int128_zero();
```
Comment (mferst) [inline target/ppc/int_helper.c:1235]: Add a comment telling that this is undefined

File: target/ppc/int_helper.c [hunk 6]
```diff
@@ -1227,6 +1227,24 @@ void helper_VDIVEUQ(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
...
+    }
+}
+
+void helper_VMODUQ(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
+{
+    if (likely(int128_nz(b->s128))) {
+        t->s128 = int128_remu(a->s128, b->s128);
+    } else {
+        t->s128 = int128_zero();
```
Comment (mferst) [inline target/ppc/int_helper.c:1244]: ditto

File: util/host-utils.c [hunk 7]
```diff
@@ -266,3 +266,182 @@ void ulshift(uint64_t *plow, uint64_t *phigh, int32_t shift, bool *overflow)
...
+        return rem;
+    }
+}
+
+/*
+ * Signed 256-by-128 division.
+ * Returns quotient via plow and phigh.
+ * Also returns the remainder via the function return value.
+ */
+Int128 divs256(Int128 *plow, Int128 *phigh, Int128 divisor)
```
[the remark this answers is not in the document]
Comment (alqotel) [reply util/host-utils.c:403]: Moved to the VDIVE commit

File: target/ppc/int_helper.c [hunk 8]
```diff
@@ -1150,6 +1150,83 @@ void helper_VDIVUQ(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
...
 
+void helper_VDIVESD(ppc_avr_t *t, ppc_avr_t *a, ppc_avr_t *b)
+{
+    int i;
+    int64_t high;
+    uint64_t low;
+    for (i = 0; i < 2; i++) {
+        high = a->s64[i];
+        low = 0;
+        if (unlikely(uabs64(a->s64[i]) >= uabs64(b->s64[i]) || !b->s64[i])) {
```
[the remark this answers is not in the document]
Comment (mferst) [reply target/ppc/int_helper.c:1161]: Ouch, you are right, this suggestion is clearly wrong

File: util/host-utils.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (alqotel) [reply util/host-utils.c:403]: Also `divs256` and `divu256` are in different commits

File: target/ppc/translate/vmx-impl.c.inc [hunk 9]
```diff
@@ -2172,6 +2172,71 @@ TRANS_VDIV_VMOD(ISA310, VDIVUD, MO_64, NULL, do_divu_i64)
...
+    /* return 0 if b = 0, so make b = 1 so the result doesn't fit in 32 bits*/
+    tcg_gen_movcond_i64(TCG_COND_EQ, val2, val2, zero, one, val2);
+
+    /* (a << 32)/b */
+    tcg_gen_shli_i64(val1, val1, 32);
+    tcg_gen_div_i64(res, val1, val2);
+
+    tcg_gen_ext32s_i64(val1, res);
+    /* if result is undefined (quotient doesn't fit in 32 bits) return 0 */
+    tcg_gen_movcond_i64(TCG_COND_EQ, res, res, val1, res, tcg_constant_i64(0));
```
Comment (mferst) [inline target/ppc/translate/vmx-impl.c.inc:2197]:
```suggestion
    tcg_gen_movcond_i64(TCG_COND_EQ, res, res, val1, res, zero);
```

**Review by mferst — APPROVED** [review]

File: target/ppc/translate/vmx-impl.c.inc [hunk 10]
```diff
@@ -3236,6 +3236,65 @@ TRANS(VMULHSD, do_vx_mulh, true , do_vx_vmulhd_i64)
...
+                   avr_full_offset(a->vrb), 16, 16, &op);               \
+                                                                        \
+    return true;                                                        \
+}
+
+    /*
+     *  If N/0 the instruction used by the backend might deliver
+     *  an invalid division signal to the process, so if b = 0 return
+     *  N/1 and if signed instruction, the same for a = int_min, b = -1
+     */
```
Comment (mferst) [inline target/ppc/translate/vmx-impl.c.inc:3261]: I guess this comment should be after the `{`

File: util/host-utils.c [hunk 11]
```diff
@@ -266,3 +266,131 @@ void ulshift(uint64_t *plow, uint64_t *phigh, int32_t shift, bool *overflow)
...
+                /* normalize the divisor, shifting the dividend accordingly */
+                divisor = int128_lshift(divisor, sh);
+                dhighest = int128_rshift(dhi, (128 - sh));
+                dhi = int128_or(int128_lshift(dhi, sh),
+                                int128_urshift(dlo, (128 - sh)));
+                dlo = int128_lshift(dlo, sh);
+
+                *phigh = udiv256_qrnnd(&dhi, dhighest, dhi, divisor);
+            } else {
+                /**
```
Comment (mferst) [inline util/host-utils.c:373]: nit: extra `*`

File: include/qemu/host-utils.h [hunk 12]
```diff
@@ -142,6 +143,19 @@ static inline int clz64(uint64_t val)
     return val ? __builtin_clzll(val) : 64;
 }
 
+/**
```
Comment (mferst) [inline include/qemu/host-utils.h:146]: nit: extra `*`
Also, the commit subject usually has a subsystem prefix. It's a bit complicated with files like host-utils, but `git blame` suggests the prefix would be `host-utils:`

Comment (mferst) [reply include/qemu/host-utils.h:146]: QEMU style shows a single `*`, I don't see why these methods would be different.

**Review by vcoracolombo — APPROVED** [review]

File: target/ppc/translate/vmx-impl.c.inc [hunk 13]
```diff
@@ -3236,6 +3236,10 @@ TRANS(VMULHSD, do_vx_mulh, true , do_vx_vmulhd_i64)
 TRANS(VMULHUW, do_vx_mulh, false, do_vx_vmulhw_i64)
 TRANS(VMULHUD, do_vx_mulh, false, do_vx_vmulhd_i64)
 
+/*
+ * This version separated the macro in 4 new macros, but kept using macros
+ */
+
```
[the remark this answers is not in the document]
Comment (alqotel) [reply target/ppc/translate/vmx-impl.c.inc:3242]: Yes, I forgot to remove yesterday after turning this version into the main one, fixed here. Thanks for the review

**Pull request closed without merging**